### ISSUE TITLE
Cannot pip_install numba

### ISSUE THREAD

**Issue by Topher-the-Geek** [bazelbuild/rules_python#415]
# 🐞 bug report

### Affected Rule

pip_install

### Is this a regression?

No

### Description

`pip_install` with the single requirement `numba` fails. It complains that it cannot find a dependent package `numpy`. It makes this complaint just after stating that it successfully installed `numpy`. It is very confused.

## 🔬 Minimal Reproduction

**Setup on Mac OS X**
```console
$ brew install llvm openssl pyenv
$ pyenv install 3.7.9
$ pyenv shell 3.7.9
```

`.bazelrc`:
```
build --action_env=CPPFLAGS="-I/usr/local/opt/llvm@9/include -I/usr/local/opt/openssl@1.1/include"
build --action_env=LDFLAGS="-L/usr/local/opt/llvm@9/lib -L/usr/local/opt/openssl@1.1/lib"
build --action_env=LLVM_CONFIG=/usr/local/opt/llvm@9/bin/llvm-config
```

`WORKSPACE`:
```
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
http_archive(
    name = "rules_python",
    url = "https://github.com/bazelbuild/rules_python/releases/download/0.1.0/rules_python-0.1.0.tar.gz",
    sha256 = "b6d46438523a3ec0f3cead544190ee13223a52f6a6765a29eae7b7cc24cc83a0",
)

register_toolchains("//:py_toolchain")

load("@rules_python//python:pip.bzl", "pip_install")

pip_install(
    name = "pip",
    requirements = "//:requirements.txt",
    timeout=1800
)
```

`BUILD`:
```
load("@pip//:requirements.bzl", "requirement")
load("@rules_python//python:defs.bzl", "py_runtime_pair")

py_runtime(
    name = "py_runtime",
    interpreter_path = "/Users/topher/.pyenv/versions/3.7.9/bin/python",
    python_version = "PY3"
)

py_runtime_pair(
    name = "py_runtime_pair",
    py2_runtime = None,
    py3_runtime = ":py_runtime",
)

toolchain(
    name = "py_toolchain",
    toolchain = ":py_runtime_pair",
    toolchain_type = "@bazel_tools//tools/python:toolchain_type",
)

py_binary(
    name = "hello",
    srcs = ["hello.py"],
    deps = [requirement("numba")]
)
```

`requirements.txt`:
```
numba
```

`hello.py`:
```
print('Hello World')
```

## 🔥 Exception or Error

```console
$ bazel build //:hello
Starting local Bazel server and connecting to it...
INFO: Repository pip instantiated at:
...
ERROR: An error occurred during the fetch of repository 'pip':
...
Successfully built llvmlite numpy
Failed to build numba
...
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "/private/var/folders/_b/rtklv78d3bgc6dx41t6l8jn00000gn/T/pip-wheel-eu4o_wgj/numba/setup.py", line 413, in <module>
      metadata['ext_modules'] = get_ext_modules()
    File "/private/var/folders/_b/rtklv78d3bgc6dx41t6l8jn00000gn/T/pip-wheel-eu4o_wgj/numba/setup.py", line 148, in get_ext_modules
      import numpy.distutils.misc_util as np_misc
  ModuleNotFoundError: No module named 'numpy'
...
```

## 🌍  Your Environment

**Operating System:**

```
macOS BigSur
11.2.1
```

**Output of `bazel version`:**

```
bazel 4.0.0-homebrew
```

**Rules_python version:**

```
0.1.0
```

**Anything else relevant?**

I can do plain old pip install at the command line fine:

```console
$ python3 -m pip install -r requirements.txt
...all goodness...
```

I see `--isolated wheel` [in extract_wheels here](https://github.com/bazelbuild/rules_python/blob/c7e068d38e/python/pip_install/extract_wheels/__init__.py#L82). If I do that at the command line, I get errors like above:

```console
$ python3 -m pip --isolated wheel -r requirements.txt
...
Building wheels for collected packages: numba, llvmlite
  Using legacy setup.py install for numba, since package 'wheel' is not installed.
  Building wheel for numba (setup.py) ... error
  ERROR: Command errored out with exit status 1:
...
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "/private/var/folders/_b/rtklv78d3bgc6dx41t6l8jn00000gn/T/pip-wheel-88k4h1wd/numba/setup.py", line 413, in <module>
      metadata['ext_modules'] = get_ext_modules()
    File "/private/var/folders/_b/rtklv78d3bgc6dx41t6l8jn00000gn/T/pip-wheel-88k4h1wd/numba/setup.py", line 148, in get_ext_modules
      import numpy.distutils.misc_util as np_misc
  ModuleNotFoundError: No module named 'numpy'
```

This is at the edge of my knowledge about Python and PIP.

I tried this with the system's python3 rather than setting up a toolchain. Oh god, THAT was a mess.

**Comment by Topher-the-Geek** [comment]
Oh wow. This is not in the bazel output, but when I run `pip wheel` at the command line, this "hiding" at the end of litany of other errors:

```
WARNING: You are using pip version 20.1.1; however, version 21.0.1 is available.
You should consider upgrading via the '/Users/topher/.pyenv/versions/3.7.9-venv02/bin/python3 -m pip install --upgrade pip' command.
```

Indeed, upgrading pip resolved the problem for running `pip wheel` at the command line. However, `bazel build` still trips over this.

I setup the toolchain for `/Users/topher/.pyenv/versions/3.7.9/bin/python` and confirmed it on the new pip:

```
$ /Users/topher/.pyenv/versions/3.7.9/bin/python -m pip -V
pip 21.0.1 from /Users/topher/.pyenv/versions/3.7.9/lib/python3.7/site-packages/pip (python 3.7)
```

But `bazel run` is getting the error like it's using the older pip. How can I get `pip_install` to use the right version of pip?

**Comment by person142** [comment]
So what's happening here is that `numba` has NumPy as a build-time dependency, and until rather recently there was no sensible way to express that for a Python package. With the introduction of `pyproject.toml` (https://www.python.org/dev/peps/pep-0518/) there is a way to do it, but it's still not widespread.

We currently hack around this by installing NumPy directly into the Python we use in our toolchain so that it's available during the execution of the repository rule-gross but it works.

In some sense the ecosystem will work itself out here as PEP 518 is more widely adopted. Until then you could potentially imagine a way to specify "extra build requirements" to `rules_python`.

**Comment by Topher-the-Geek** [comment]
Yeah, it's a hack, but one that improves things. Prior to isolating this issue, I was contemplating putting ALL my requirements into my base Docker image, and loosing all of Bazel's goodness for creating things with just the right dependencies. Now I need put only `numpy` (and its dependencies) into my base image.

**Comment by groodt** [comment]
@Topher-the-Geek Try using one of the new commits. https://github.com/bazelbuild/rules_python/commits/master

The 0.1.0 tag is a bit old now.

There is a newer PR waiting to bump pip to the latest, however, it drops support for Python 2 and Python 3.5, so it probably needs to have an official release tag before it's merged.

**Comment by Topher-the-Geek** [comment]
Bliss! Both of these work:

Latest commit on master

```
rules_python_version = "c7e068d38e2fec1d899e1c150e372f205c220e27"
http_archive(
    name = "rules_python",
    url = "https://github.com/bazelbuild/rules_python/archive/{}.zip".format(rules_python_version),
    strip_prefix = "rules_python-{}".format(rules_python_version),
    sha256 = "4b6055b232bfdef4cbde1ea620a478c57b032e96527da3472c2b49dbc1e42048"
)
```

Latest commit on PR #412 

```
rules_python_version = "b12e6e89b2b949aacc9c2bdc27de4f92523d2f5a"
http_archive(
    name = "rules_python",
    url = "https://github.com/bazelbuild/rules_python/archive/{}.zip".format(rules_python_version),
    strip_prefix = "rules_python-{}".format(rules_python_version),
    sha256 = "7b72f06540bb323b711533196c338b2a5016653c975612e5137925da6e396500"
)
```

I see the PR is failing on Ubuntu 16.04. I don't much care about that. We are on BigSur for local dev, and Ubuntu 20 for cloud dev and prod.

Since a fix is forthcoming, it seems reasonable to close this issue. I'll leave that up to you all depending on your workflow.

**Comment by hrfuller** [comment]
The takeaway from this ticket is we should likely do a release with the new pip depenencies. @thundergolfer shall we?

**Comment by thundergolfer** [comment]
@Topher-the-Geek with release `0.2.0` I can successfully install `numba==0.53.1` (latest) and run a `py_binary` using it. 

I think this resolves the issue, but let me know if there's anything else to address.
